### ISSUE TITLE
Use patch_overlap in patch based inference

### ISSUE THREAD

**Issue by GFabien** [fepegar/torchio#181]
Hi @fepegar I have trouble with the use of the `patch_overlap` attribute in the patch based inference pipeline. Indeed, reading [NiftyNet tutorial](https://niftynet.readthedocs.io/en/dev/window_sizes.html) I thought I did not need to use `patch_overlap` (i.e. keep it to 0). 

However, after some experiments I've seen that I get much better results using a non zero `patch_overlap` (4 for example) which seems intuitive as it should be harder to predict classes in the patches' borders as less context information is available.

That would have been fine is this was the end of the story but I discovered something else. Indeed, visualizing the predictions with the different values of `patch_overlap` I realized that when I had a non zero value I got all the borders of my whole volume that were totally wrong (value of 0.5 in my case while background should be 0 and grey matter 1). Quantitatively it results in a increase of Dice Loss of more than 0.1 compared to the same prediction with a crop of the border of the volume. After diving into the code I have the impression that all patches are aggregated by cropping them to remove overlapping voxels but no special action is made for border patches so it takes the default 0 value set when `_output_tensor ` is created.

Should there be a specific handling of patches on the border or is there something I missed with the use of the `patch_overlap` variable? Thanks for your help!

**Comment by fepegar** [comment]
I'm not sure I follow. Do you have some visualization or some code to reproduce? Is this maybe an issue related to [volume padding before inference](https://niftynet.readthedocs.io/en/dev/config_spec.html#volume-padding-size)?

**Comment by GFabien** [comment]
> Do you have some visualization or some code to reproduce?

Here are two inferences from the same model, the left one is made with a `patch_overlap` of 0 while the right one is made using a `patch_overlap` of 4.
![wrong_border](https://user-images.githubusercontent.com/39873986/83742017-2cac2200-a659-11ea-88d3-617eac3dba3d.png)
You can see on the right one a thin greenish border around the volume which comes from the `GridAggregator._output_tensor` but not from the predictions of the model.

> Is this maybe an issue related to volume padding before inference?

It might be, is this something we should handle manually?

**Comment by fepegar** [comment]
Thanks for the screenshot. I'll look into this tomorrow (I'm trying to work on the library only on Fridays, I have a PhD to do 😅).

In the meantime, I recommend you stop using jet (I guess here)
https://github.com/romainVala/torchQC/blob/56e46f43c88a12a1885edce3646dff47d448a3dd/plot_dataset.py#L242

and switch to a [perceptually uniform colormap](https://bids.github.io/colormap/). I like using [diverging colormaps](https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html#diverging) for probabilities because white represents voxels where the image is uncertain.

**Comment by romainVala** [comment]
> Thanks for the screenshot. I'll look into this tomorrow (I'm trying to work on the library only on Fridays, I have a PhD to do sweat_smile).

come on ! that library could become more important than your phd  (for sure more usefull)
just joking, I understand, I hope you still manage both

**Comment by fepegar** [comment]
> come on ! that library could become more important than your phd (for sure more usefull)

😂 that's an open secret! Thanks Romain!

**Comment by GFabien** [comment]
> Thanks for the screenshot. I'll look into this tomorrow (I'm trying to work on the library only on Fridays, I have a PhD to do 😅).

Haha yes I understand! If you need help regarding some issues I'll be glad to give a hand!

**Comment by GFabien** [comment]
FYI I investigated the issue and changing the `GridAggregator` `add_batch` method solves my problem but I think the best solution would be to pad the image by an amount of `patch_overlap` and crop the output_tensor back to the original shape when calling `get_output_tensor`.

My quick and very dirty fix looks like that:
```python
    def add_batch(self, patches: torch.Tensor, locations: TypeData) -> None:
        max_i, max_j, max_k = self._output_tensor.shape[1:]
        patches = patches.cpu()
        self._ensure_output_dtype(patches)

        for patch, location in zip(patches, locations):
            i_ini, j_ini, k_ini, i_fin, j_fin, k_fin = location
            border = np.array([*self.patch_overlap, *-np.array(self.patch_overlap)])
            if i_ini == 0:
                border[0] = 0
            if j_ini == 0:
                border[1] = 0
            if k_ini == 0:
                border[2] = 0
            if i_fin + 1 == max_i:
                border[3] = 0
            if j_fin + 1 == max_j:
                border[4] = 0
            if k_fin + 1 == max_k:
                border[5] = 0

            i_ini, j_ini, k_ini, i_fin, j_fin, k_fin = location + border
            bi_ini, bj_ini, bk_ini, bi_fin, bj_fin, bk_fin = map(lambda x: None if x == 0 else x, border)
            patch = patch[
                :,  # channels dimension
                bi_ini:-bi_fin,
                bj_ini:-bj_fin,
                bk_ini:-bk_fin,
            ]

            channels = len(patch)
            for channel in range(channels):
                self._output_tensor[channel, i_ini:i_fin, j_ini:j_fin, k_ini:k_fin] = patch[channel]
```

**Comment by fepegar** [comment]
Thank you! This will definitely help me understand.

**Comment by fepegar** [comment]
So I've started to look at these classes and I really want to refactor them, because it's hard to know what's going on. They are from the early days of this library and almost pasted from NiftyNet. Also, I stopped doing patch-based training so I haven't paid much attention to samplers and aggregators (until now).

For now, I think that the grid sampler is not working properly:

```python
import torch
import torchio

image_size = 1, 1, 5
patch_size = 1, 1, 3
subject = torchio.Subject(
    img=torchio.Image(tensor=torch.rand(image_size))
)
sample = torchio.ImagesDataset([subject])[0]
sampler = torchio.data.GridSampler(
    sample,
    patch_size,
    0,
)
print(sampler.locations)
```

Output:

```python
[[0 0 0 1 1 3]
 [0 0 2 1 1 5]
 [0 0 1 1 1 4]]
```

I think the last location shouldn't be there.


This is the output (only initial voxel) for an image of size 10 and different patch sizes:

```python
Patch size: 1
[[0 0 0]
 [0 0 1]
 [0 0 2]
 [0 0 3]
 [0 0 4]
 [0 0 5]
 [0 0 6]
 [0 0 7]
 [0 0 8]
 [0 0 9]]

Patch size: 2
[[0 0 0]
 [0 0 2]
 [0 0 4]
 [0 0 6]
 [0 0 8]]

Patch size: 3
[[0 0 0]
 [0 0 3]
 [0 0 6]
 [0 0 7]]

Patch size: 4
[[0 0 0]
 [0 0 4]
 [0 0 6]]

Patch size: 5
[[0 0 0]
 [0 0 5]
 [0 0 2]]

Patch size: 6
[[0 0 0]
 [0 0 4]
 [0 0 2]]

Patch size: 7
[[0 0 0]
 [0 0 3]
 [0 0 2]]

Patch size: 8
[[0 0 0]
 [0 0 2]
 [0 0 1]]

Patch size: 9
[[0 0 0]
 [0 0 1]]

Patch size: 10
[[0 0 0]]
```

Here, the locations seem wrong for patch sizes 5, 6, 7 and 8. Weird.

**Comment by GFabien** [comment]
Yes I agree that the locations you pointed out are strange, I will have a look too

**Comment by fepegar** [comment]
For now, I think I have a much more compact way of getting the indices:

```python
def get_grid_locations(image_size, patch_size, patch_overlap):
    indices = []

    zipped = zip(image_size, patch_size, patch_overlap)
    for im_size_dim, patch_size_dim, patch_overlap_dim in zipped:
        end = im_size_dim + 1 - patch_size_dim
        step = patch_size_dim - 2 * patch_overlap_dim
        indices_dim = list(range(0, end, step))
        if im_size_dim % step:
            indices_dim.append(im_size_dim - patch_size_dim)
        indices.append(indices_dim)
    locations = np.array(np.meshgrid(*indices)).reshape(3, -1).T
    return locations
```

**Comment by fepegar** [comment]
> FYI I investigated the issue and changing the GridAggregator add_batch method solves my problem but I think the best solution would be to pad the image by an amount of patch_overlap and crop the output_tensor back to the original shape when calling get_output_tensor.
> My quick and very dirty fix looks like that

So if I understand correctly, you¡re not cropping patches at the border of the volume. This seems like a good solution to me, I think it's easier to implement than padding and cropping, and makes sense conceptually.

**Comment by romainVala** [comment]
yes, make sense
